### PR TITLE
251 cicd review and enhancement for didcomm mediator project e3

### PR DESCRIPTION
**Pull request by ndefokou**
@Christiantyemele @Hermann-Core @IngridPuppet Please review this PR

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -1,13 +1,20 @@
...
+name: Rust CI/CD
 
-on: [push, pull_request]
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
```
Comment (IngridPuppet) [inline .github/workflows/rust.yml:9]: Create another workflow file for CD, the triggers  are generally not the same. You run CI on every open pull request, and you run CD on merge. You would not want to run CD as soon as someone creates a PR.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -35,14 +42,48 @@ jobs:
...
+        HOST: ${{ secrets.PRODUCTION_HOST }}
+        USER: ${{ secrets.PRODUCTION_USER }}
       run: |
-        cargo build --workspace --all-features
-        cargo nextest run --workspace --all-features
+        echo "${{ secrets.SSH_PRIVATE_KEY }}" > ssh_key
+        chmod 600 ssh_key
+        scp -i ssh_key target/release/your_binary ${{ env.USER }}@${{ env.HOST }}:/path/to/your/binary
+        ssh -i ssh_key ${{ env.USER }}@${{ env.HOST }} 'systemctl restart your-service'
+        rm -f ssh_key
```
Comment (IngridPuppet) [inline .github/workflows/rust.yml:89]: Where are you deploying the server to? Aren't we supposed to have an AWS instance set up for the project? I don't get the logic here...

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -1,13 +1,20 @@
...
 
 env:
   CARGO_TERM_COLOR: always
 
 jobs:
-  build:
-    name: Build and Test
+  # Continuous Integration Job
+  ci:
+    name: Continuous Integration
```
Comment (Hermann-Core) [inline .github/workflows/rust.yml:17]: Why do you change the name here? Keep the previous name. CI means nothing, it is a process not a step

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -35,14 +42,48 @@ jobs:
         rustflags:
 
     - name: Install Nextest test runner
-      uses: taiki-e/install-action@nextest
+      run: cargo install cargo-nextest
```
Comment (Hermann-Core) [inline .github/workflows/rust.yml:45]: Why did you remove the previous step command ?

File: .github/workflows/rust.yml [hunk 5]
```diff
@@ -35,14 +42,48 @@ jobs:
...
+
+    - name: Build Project
+      run: cargo build --workspace --all-features
+
+    - name: Run Tests
+      run: cargo nextest run --workspace --all-features
+
+  # Continuous Deployment Job
+  cd:
+    name: Continuous Deployment
```
Comment (Hermann-Core) [inline .github/workflows/rust.yml:63]: Again this name means nothing. And as Ingrid suggested, move the CD processes to a different file.

**Review by Hermann-Core — CHANGES_REQUESTED** [review]

File: .github/workflows/rust.yml (hunk 2 above)
Comment (ndefokou) [reply .github/workflows/rust.yml:89]: We are to Have

**Comment by ndefokou** [comment]
@IngridPuppet @IngridPuppet You can check the CI for the CD; we are going to discuss it with Mr. Francis in our meeting.

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -1,13 +1,20 @@
 name: Rust CI
 
-on: [push, pull_request]
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
```
Comment (IngridPuppet) [inline .github/workflows/CI.yml:9]: Triggering the workflow on both push and pull_request events makes it run twice unnecessarily, when a PR is open, and also when it is merged. Additionally, all PRs, even when not directed to main, should be subject to these checks.

```suggestion
on: [pull_request]
```

File: .github/workflows/CD.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+name: CD 
+
+on:
+  workflow_run:
+    workflows: ["Rust CI Pipeline"]  # Trigger on successful CI completion
+    types:
+      - completed
```
Comment (IngridPuppet) [inline .github/workflows/CD.yml:7]: A successful CI pipeline is still not sufficient to deploy anything. CD should be triggered when a PR is merged, which fires a push event to main.

```suggestion
on:
  push:
    branches: [main]
```

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml (hunk 6 above)
Comment (ndefokou) [reply .github/workflows/CI.yml:9]: Okay thanks let me change that

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -1,13 +1,14 @@
 name: Rust CI
 
-on: [push, pull_request]
+on:[push, pull_request]
```
Comment (IngridPuppet) [inline .github/workflows/CI.yml:3]: Is there a specific reason why you want to keep the push trigger here? Note that it places an unnecessary stress on pipeline resources, especially when all branches are eligible.

```suggestion
on: [pull_request]
```

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

**Review by IngridPuppet — APPROVED** [review]

File: .github/workflows/CI.yml (hunk 8 above)
Comment (ndefokou) [reply .github/workflows/CI.yml]: From the other PR i am seeing that the CI is done twice and i think it is because of that last CI 
It is now clear to me

**Review by IngridPuppet — APPROVED** [review]

File: .github/workflows/CD.yml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  push:
+    branches: [main]
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
Comment (Hermann-Core) [inline .github/workflows/CD.yml:13]: I think there's a recent version of this action.

File: .github/workflows/CI.yml [hunk 10]
```diff
@@ -35,14 +36,18 @@ jobs:
         rustflags:
 
     - name: Install Nextest test runner
-      uses: taiki-e/install-action@nextest
+      run: cargo install cargo-nextest
```
Comment (Hermann-Core) [inline .github/workflows/CI.yml:39]: Why did you remove the previous action? The one you've replaced it with will execute at every CI pipeline run, which is not what we want.

**Review by Hermann-Core — CHANGES_REQUESTED** [review]

File: .github/workflows/CI.yml (hunk 10 above)
Comment (ndefokou) [reply .github/workflows/CI.yml]: Okay i see .Let me fix that

File: .github/workflows/CD.yml (hunk 9 above)
Comment (Hermann-Core) [reply .github/workflows/CD.yml:13]: Of course there’s is a version 4. Make sure to use the latest versions of the actions you’re using.

**Review by Hermann-Core — APPROVED** [review]
